dvc/remote/azure.py

```diff
@@ -0,0 +1,241 @@
+            raise ValueError('Azure Storage connection string missing')
+
+        self.__blob_service = None
+
+    @property
+    def blob_service(self):
+        if self.__blob_service is None:
+            self.__blob_service = BlockBlobService(
+                connection_string=self.connection_string)
+            self.__blob_service.create_container(self.bucket)
+        return self.__blob_service
+
+    def _get_etag(self, bucket, key):
+        try:
+            blob = self.blob_service.get_blob_properties(bucket, key)
+            return blob.properties.etag
+        except Exception:
+            return None
+
+    def save_info(self, path_info):
+        if path_info['scheme'] != self.scheme:
+            raise NotImplementedError
+
+        return {self.PARAM_ETAG: self._get_etag(
+            path_info['bucket'], path_info['key'])}
+
+    def save(self, path_info):
+        if path_info['scheme'] != self.scheme:
+            raise NotImplementedError
+
+        etag = self._get_etag(path_info['bucket'], path_info['key'])
+        dest_key = '{}/{}'.format(etag[0:2], etag[2:])
+
+        self._copy_blob(to_bucket=self.bucket,
+                        to_key=dest_key,
+                        from_bucket=path_info['bucket'],
+                        from_key=path_info['key'])
+
+        return {self.PARAM_ETAG: etag}
+
+    def _copy_blob(self, to_bucket, to_key, from_bucket, from_key):
+        source = self.blob_service.make_blob_url(from_bucket, from_key)
+
+        copy = self.blob_service.copy_blob(to_bucket, to_key, source)
+
+        if self.COPY_POLL_SECONDS <= 0:
+            return
+
+        while copy.status != 'success':
+            time.sleep(self.COPY_POLL_SECONDS)
+            copy = self.blob_service.get_blob_properties(
+                to_bucket, to_key).properties.copy
+
+    def checkout(self, path_info, checksum_info):
+        if path_info['scheme'] != self.scheme:
+            raise NotImplementedError
+
+        etag = checksum_info.get(self.PARAM_ETAG, None)
+        if not etag:
+            return
+
+        self._copy_blob(to_bucket=path_info['bucket'],
+                        to_key=path_info['key'],
+                        from_bucket=self.bucket,
+                        from_key='{}/{}'.format(etag[0:2], etag[2:]))
+
+    def remove(self, path_info):
+        if path_info['scheme'] != self.scheme:
+            raise NotImplementedError
+
+        Logger.debug('Removing azure://{}/{}'.format(path_info['bucket'],
+                                                     path_info['key']))
+
+        self.blob_service.delete_blob(path_info['bucket'], path_info['key'])
+
+    def md5s_to_path_infos(self, md5s):
+        return [{
+            'scheme': self.scheme,
+            'bucket': self.bucket,
+            'key': '{}/{}'.format(md5[0:2], md5[2:])
+        } for md5 in md5s]
+
+    def exists(self, path_infos):
+        # TODO: why does S3 fetch all blobs first?
```

Comment:
It is simply much faster :) We mostly use exists() method when filtering a bulk of cache files to decide if we need to download/upload them and in s3 list_objects is much-much faster than trying to check keys one-by-one. If blob_service.exists() method is plenty fast as is, there is no reason to redo this part of code.

Reply:
Let's put this as a comment instead of this TODO.